### PR TITLE
Moved shared experience mechanism to lua

### PR DESCRIPTION
**Pull request by edumntg**
Moved the shared experience mechanism to events and also the shared
experience formula was fixed.

### CODE REVIEW

File: data/events/scripts/party.lua [hunk 1]
```diff
@@ -9,3 +9,24 @@ end
 function Party:onDisband()
 	return true
 end
+
+function Party:onShareExperience(exp)
+	local sharedExperienceMultiplier = 1.20 --20%
+	local vocationsIds = {}
+	table.insert(vocationsIds, self:getLeader():getVocation():getId())
```
Comment (ninjalulz) [inline data/events/scripts/party.lua]: There should be a if statement around this (it shouldn't insert a vocationId if it's equal to VOCATION_NONE), and insert base vocationId instead (same goes for members).

File: data/events/scripts/party.lua [hunk 2]
```diff
@@ -9,3 +9,24 @@ end
...
+	for _, member in ipairs(self:getMembers()) do
+		if not isInArray(vocationsIds, member:getVocation():getId()) then
+			table.insert(vocationsIds, member:getVocation():getId())
+		end
+	end
+	
+	local size = #vocationsIds
+	if size > 1 then
+		sharedExperienceMultiplier = 1.0 + ((size * (10 + (size - 1) * 5)) / 100)
+	else
```
Comment (ninjalulz) [inline data/events/scripts/party.lua]: You can remove L26-27 (you already assigned the value 1.20 to the variable when you declared it).

Comment (edumntg) [inline data/events/scripts/party.lua]: It's true, i just copied the code from sources and changed it to lua.

**Comment by ranisalt** [comment]
I'd really like to see this implemented on a huge server to check if this does not bottleneck the server.

File: src/events.cpp [hunk 3]
```diff
@@ -337,6 +340,39 @@ bool Events::eventPartyOnDisband(Party* party)
...
+
+	LuaScriptInterface::pushUserdata<Party>(L, party);
+	LuaScriptInterface::setMetatable(L, -1, "Party");
+
+	lua_pushnumber(L, exp);
+
+	if (scriptInterface.protectedCall(L, 2, 1) != 0) {
+		LuaScriptInterface::reportError(nullptr, LuaScriptInterface::popString(L));
+	}
+	else {
```
Comment (WibbenZ) [inline src/events.cpp]: } else {

File: data/events/scripts/party.lua [hunk 4]
```diff
@@ -9,3 +9,29 @@ end
...
+	local vocationsIds = {}
+	
+	local vocationId = self:getLeader():getVocation():getId()
+	if vocationId ~= VOCATION_NONE then
+		table.insert(vocationsIds, vocationId)
+	end
+	
+	for _, member in ipairs(self:getMembers()) do
+		vocationId = member:getVocation():getId()
+		if not isInArray(vocationsIds, vocationId) and vocationId ~= VOCATION_NONE then
```
Comment (WibbenZ) [inline data/events/scripts/party.lua]: https://github.com/otland/forgottenserver/blob/master/data/lib/compat/compat.lua#L1054

File: data/events/scripts/party.lua [hunk 5]
```diff
@@ -9,3 +9,29 @@ end
 function Party:onDisband()
 	return true
 end
+
+function Party:onShareExperience(exp)
+	local sharedExperienceMultiplier = 1.20 --20%
+
```
Comment (WibbenZ) [inline data/events/scripts/party.lua]: Remove all the trailing spaces

File: src/party.cpp [hunk 6]
```diff
@@ -418,7 +390,9 @@ bool Party::setSharedExperience(Player* player, bool _sharedExpActive)
 
 void Party::shareExperience(uint64_t experience, Creature* source/* = nullptr*/)
 {
-	uint32_t shareExperience = static_cast<uint64_t>(std::ceil(((static_cast<double>(experience) / (memberList.size() + 1)) + (static_cast<double>(experience) * extraExpRate))));
+	uint64_t shareExperience = experience;
+	g_events->eventPartyOnShareExperience(this, shareExperience);
+
```
Comment (WibbenZ) [inline src/party.cpp]: Remove trailing

**Comment by kito2** [comment]
Great idea!

**Pull request closed without merging**